test/DecryptionFailureTracker-test.js

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2018 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import expect from 'expect';
+
+import DecryptionFailureTracker from '../src/DecryptionFailureTracker';
+
+import { MatrixEvent } from 'matrix-js-sdk';
+
+function createFailedDecryptionEvent() {
+    const event = new MatrixEvent({
+        event_id: "event-id-" + Math.random().toString(16).slice(2),
+    });
+    event._setClearData(
+        event._badEncryptedMessage(":("),
+    );
+    return event;
+}
+
+describe.only('DecryptionFailureTracker', function() {
+    it('tracks a failed decryption', function(done) {
+        const failedDecryptionEvent = createFailedDecryptionEvent();
+        let trackedFailure = null;
+        const tracker = new DecryptionFailureTracker((failure) => {
+            trackedFailure = failure;
+        });
+
+        tracker.eventDecrypted(failedDecryptionEvent);
+
+        // Pretend "now" is Infinity
+        tracker.checkFailures(Infinity);
```

Comment:
Oh, I see why this got passed in. We have mock clocks for tests, right?

Reply:
Yep, but the less mocking, the better I think.

Reply:
To an extent, although there is also an argument for not contorting the API for the sake of the tests.

Reply:
Contorting the API is an unfair assessment. Bearing in mind, if we mocked the clocks, the tests would do more and more functions in the class would have to care about `Date.now()`. Really not seeing the advantage of that.

Reply:
Mostly what I saw was the start method having to make two anonymous functions to go & call Date.now and then pass it into the function, when the function could have called it itself. I agree though it does make the tests easier to write, although it still has to leave the DFT to call `Date.now()` when `eventDecrypted`, but then pass in the value for `Date.now()` when calling trackFailure, so this is a bit of a mismatch.

Not that I really care about any of this enough to push for changing it, so this is fine.

Reply:
> it still has to leave the DFT to call Date.now() when eventDecrypted, but then pass in the value for Date.now() when calling trackFailure

true that half of the API needs you to pass in a Date.now(). We could equally ask that instead of sending matrix events into the class, send DecryptionFailures instead but the fact that the DecryptionFailure is hidden away seemed to be quite a nice property.

>Not that I really care about any of this

Well I care and I value your point of view :)